treeherder/auth/backends.py

```diff
@@ -37,48 +36,50 @@ class TaskclusterAuthBackend(object):
          'expires': '2016-10-31T17:40:45.692Z'}
     """
 
-    def _get_scope_value(self, scopes, scope_prefix):
-        for scope in scopes:
-            if scope.startswith(scope_prefix):
-                return scope[len(scope_prefix):]
-        return None
-
-    def _get_email(self, result):
+    def _extract_email_from_clientid(self, client_id):
         """
-        Get the user's email from the mozilla-user scope
+        Extract the user's email from the client_id
 
-        For more info on scopes:
-        https://docs.taskcluster.net/manual/3rdparty#authenticating-with-scopes
+        The client_id MUST be in the form:
+            "email/<username>@<domain>"
+              or
+            "mozilla-ldap/<username>@<domain>"
         """
-        email = self._get_scope_value(result["scopes"], "assume:mozilla-user:")
-
-        if email and re.search(r'.+@.+', email):
-            return email
-        raise TaskclusterAuthenticationFailed(
-            "Invalid email for clientId: '{}'. Invalid value for scope 'assume:mozilla-user': '{}'".format(
-                result["clientId"], email))
 
-    def _get_user(self, email):
+        if client_id.startswith("email/") or client_id.startswith("mozilla-ldap/"):
+            email = client_id.split("/", 1)[1]
+            # ensure what we find after the split is a valid email and ONLY
+            # a valid email
+            match = re.search(
+                r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)",
+                email)
+            if match:
+                return match.group(0)
+        raise NoEmailException("No email found in clientId: '{}'".format(client_id))
```

Comment:
As much as I don't like over-complex regexes, we have to have the email regex anyway, so another way of writing this might be to just do it all in one:
```python
# Client IDs must be in one of these forms:
# - email/foo@bar.com
# - mozilla-ldap/foo@bar.com
# Email regex taken from http://emailregex.com
CLIENT_ID_RE = re.compile(r"^(?:email|mozilla-ldap)/([a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+)$")

match = CLIENT_ID_RE.match(client_id)
if match:
    return match.group(1)

raise NoEmailException("No email found in clientId: '{}'".format(client_id))
```

(With the `CLIENT_ID_RE` defined perhaps on the module rather than in this function)

Though I'm happy for you to pick which you prefer :-)